runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -6482,6 +6482,53 @@ TR::Register *J9::Power::TreeEvaluator::VMnewEvaluator(TR::Node *node, TR::CodeG
             genInitArrayHeader(node, iCursor, isVariableLen, clazz, NULL, resReg, zeroReg, condReg, enumReg, dataSizeReg,
                   tmp5Reg, tmp4Reg, conditions, needZeroInit, cg);
 
+#ifdef TR_TARGET_64BIT
+         TR::Register *offsetReg = tmp5Reg;
+         TR::Register *firstDataElementReg = tmp4Reg;
+         TR::MemoryReference *dataAddrSlotMR = NULL;
+
+         if (TR::Compiler->om.compressObjectReferences() && isVariableLen)
+            {
+            // We need to check sizeReg at runtime to determine correct offset of dataAddr field.
+            if (comp->getOption(TR_TraceCG))
+               traceMsg(comp, "Node (%p): Dealing with compressed refs variable length array.\n", node);
+
+            TR_ASSERT_FATAL_WITH_NODE(node, (fej9->getOffsetOfDiscontiguousDataAddrField() - fej9->getOffsetOfContiguousDataAddrField()) == 8, "DataAddr field offset for discontiguous arrays is expected to be 8 bytes more than dataAddr field offset for contiguous arrays. But was %d bytes for discontigous and %d bytes for contiguous.\n", fej9->getOffsetOfDiscontiguousDataAddrField(), fej9->getOffsetOfContiguousDataAddrField());
+
+            iCursor = generateTrg1Src1Instruction(cg, TR::InstOpCode::neg, node, offsetReg, enumReg);
+            iCursor = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::sradi, node, offsetReg, offsetReg, 63);
+            // andi_r works because sradi replicates bit 0 of offsetReg to fill the vacated positions. 1 for length > 0 and 0 otherwise.
+            iCursor = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::andi_r, node, offsetReg, offsetReg, 8);
+            iCursor = generateTrg1Src1Instruction(cg, TR::InstOpCode::neg, node, offsetReg, offsetReg);
+            iCursor = generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, offsetReg, resReg, offsetReg);
```

Comment:
first of all, i believed we never allocated inline truly discontiguous array (at least i am not aware of ... and the size is over the default TLH size to begin with).  I am wondering if you need to deal with it really.

secondly, if you need to deal with it, by my quick reading of it, i estimated you can do it in 3 instructions at most instead of 5:  trying to get 0 or 8 into offsetReg according to sign of enumReg, then do it contiguous way.

Reply:
> first of all, i believed we never allocated inline truly discontiguous array (at least i am not aware of ... and the size is over the default TLH size to begin with). I am wondering if you need to deal with it really.

Is this a Power codegen statement only? We certainly do inline allocations of 0 length arrays on x86 and Z as they are quite common. If we are reverting to a VM helper to allocate 0 length arrays then there is definitely a performance opportunity to capture here. For example on DayTrader zero-length array allocations are reasonably common.

Reply:
> Is this a Power codegen statement only?

no, i didn't mean it for P only, but i expected 0-length (or shorter than a certain amount) array is in contiguous form. i have vague impression that "certain amount" is bigger than the default max TLH (128KB).

Reply:
The size field in contiguous array header is at the same offset as `mustBeZero` field of discontiguous array. We first check for size at object + 4 bytes. if that field is 0 the array could be either a discontiguous array or a zero length array. To be sure we need to check another field at object + 8 bytes.
```c++
typedef struct J9IndexableObjectDiscontiguousCompressed {
	U_32 clazz;
	U_32 mustBeZero;
	U_32 size;
	U_32 padding;
#if defined(J9VM_ENV_DATA64)
	void *dataAddr;
#endif /* J9VM_ENV_DATA64 */
} J9IndexableObjectDiscontiguousCompressed;
```
```c++
typedef struct J9IndexableObjectContiguousCompressed {
	U_32 clazz;
	U_32 size;
#if defined(J9VM_ENV_DATA64)
	void *dataAddr;
#endif /* J9VM_ENV_DATA64 */
} J9IndexableObjectContiguousCompressed;
```

Reply:
as far as i know, enumReg contains the true array length, not the potentially faked length loaded from object (depending on whether it is discontiguous or contiguous).  you still need to think through whether non-zero length discontiguous array allocation can come here.  then, decide how to use enumReg in dataAddr set-up.

Reply:
I believe `dataSizeReg` is what contains the true array length. We evaluate first child of the new array node into `enumReg` [1] and never modify it throughout the evaluator. Only other time `enumReg` might have a different value is [2], but that path is taking only for `TR::new`.

As for the instruction sequence, I have updated it according to your suggestions from the call last Friday. For context, the concern with the current sequence was that it will only work with 63 bit integers. It will fail if `enumReg` has 1 in the 0th bit (sign bit) because we rely on 0th bit to always be 0.

[1] https://github.com/eclipse-openj9/openj9/blob/7ff101e482a05d434dbc1108051a024fea39b0f0/runtime/compiler/p/codegen/J9TreeEvaluator.cpp#L6375 "Evaluate first child into enumReg"
[2] https://github.com/eclipse-openj9/openj9/blob/7ff101e482a05d434dbc1108051a024fea39b0f0/runtime/compiler/p/codegen/J9TreeEvaluator.cpp#L6411-L6412 "Alternate path for enumReg"

Reply:
Ambiguous "array length" is used throughout here:  enumReg is the true arrayObject.length(), whereas dataSizeReg is the length in byte.